blockchain/blockindex.go

```diff
@@ -574,14 +882,424 @@ func (bi *blockIndex) SetStatusFlags(node *blockNode, flags blockStatus) {
+	// now eligible for validation when the parent of the passed block is also
+	// eligible (or has already been validated).
+	var linkedBlocks []*blockNode
+	bi.Lock()
+	if bi.canValidate(node.parent) {
+		linkedBlocks = bi.linkBlockData(node, tip)
+	} else if !node.parent.status.KnownInvalid() {
+		unlinkedChildren := bi.unlinkedChildrenOf[node.parent]
+		bi.unlinkedChildrenOf[node.parent] = append(unlinkedChildren, node)
+	}
+	bi.Unlock()
+	return linkedBlocks
+}
+
+// workSorter implements sort.Interface to allow a slice of block nodes to be
+// sorted according to the cumulative proof of work, then block availability,
+// then the order in which the data was received, and finally, based on the hash
+// in order to ensure a stable sort order.
+//
+// The sort order is ascending, meaning the final item is the best candidate for
+// the purposes of best chain selection.
+//
+// It should be noted that it relies on fields in the block node that are
+// protected by the block index mutex, so sorting must not be performed without
+// holding it.
+type workSorter []*blockNode
+
+// Len returns the number of block nodes in the slice.  It is part of the
+// sort.Interface implementation.
+func (s workSorter) Len() int {
+	return len(s)
+}
+
+// Swap swaps the block nodes at the passed indices.  It is part of the
+// sort.Interface implementation.
+func (s workSorter) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+// Less returns whether the node with index i should sort before the node with
+// index j.  It is part of the sort.Interface implementation.
+//
+// This function MUST be called with the block index lock held (for reads).
+func (s workSorter) Less(i, j int) bool {
+	return workSorterLess(s[i], s[j])
+}
+
+// FindBestChainCandidate searches the block index for the best potentially
+// valid chain that contains the most cumulative work and returns its tip.  In
+// order to be potentially valid, all of the block data leading up to a block
+// must have already been received and must not be part of a chain that is
+// already known to be invalid.  A chain that has not yet been fully validated,
+// such as a side chain that has never been the main chain, is neither known to
+// be valid nor invalid, so it is possible that the returned candidate will form
+// a chain that is invalid.
+//
+// This function is safe for concurrent access.
+func (bi *blockIndex) FindBestChainCandidate() *blockNode {
+	bi.RLock()
+	defer bi.RUnlock()
+
+	numCandidates := len(bi.bestChainCandidates)
+	switch {
+	case numCandidates == 1:
+		// Nothing more to do when there is only a single candidate.
+		for node := range bi.bestChainCandidates {
+			return node
+		}
+
+	case numCandidates > 1:
+		// Sort potential best chain candidates in descending order according to
+		// their cumulative proof of work so the highest work candidates are
+		// first.
+		sortedCandidates := make([]*blockNode, 0, numCandidates)
```

Comment:
Upon reviewing this again, I see that this is a bit wasteful.  There is really no reason to allocate a slice and perform a full sort it here because only a single entry is ever returned.  Instead, it would be more efficient to simply iterate the candidates and find the best one.

Further, this is the only place that the sorting is done, so if it is changed, the `workSorter` implementation to support `sort.Interface` for the `sort.Sort` call would no longer be needed allowing the code to be removed.

Reply:
Done.